app/models/split.rb

```diff
@@ -82,8 +119,34 @@ def create_decision!(params = {})
     build_decision(params).tap(&:save!)
   end
 
+  def registry
+    if try(:feature_incomplete?)
+      knock_out_weightings(super)
+    else
+      super
+    end
+  end
+
   private
 
+  def knock_out_weightings(registry_hash, to: "false")
+    target_variant = to.to_s
+    knock_out_weightings_if_possible(registry_hash, to: target_variant) || log_knockout_error(target_variant) && registry_hash
```

Comment:
Nitpicking on the control flow here a little. It's very compact, but really what's happening is that this:

```
f(A) || log_error && A
```

Is doing some immediate-turnaround nil-handling for `f`, whose last line looks like

```
found_key ? A' : nil
```

When combined, those could look more like this:

```
if found_key
  A'
else
  log_error
  A
end
```

To me, that groks a bit more immediately, and also has the nice property of not caring about the return value of `log_error`.

Reply:
A more rubyesque version of that last conditional could look more like this:

```ruby
f(A) || A
```

Where `f` looks like this:

```ruby
found_key? A' : log_error && nil
```

In my mind that does lose a little grokability again, but it might be a nice middle-ground between the existing, compact control flow, and a variation that doesn't care about whether log_error returns `true`

Reply:
Yeah, I was threading a needle of rubocop line length here without adding a class, I had all those factoring along the way iirc. Note that your last suggestion would need a second arg to f, it’s probably an object envy issue if I really wanna make it work nice. Should I do that?

Reply:
Feels _almost_ like `registry` wants to deserialize to a value object that can perform all of this knock-out magic internally. (Some kind of hash-like that accepts the additional context of `feature_incomplete` and whatever provides the `to` value.) But that said, I'm leaning more towards leaving it as-is (as you said, this code is not going to need to change much), unless something about that idea resonates with you.

Reply:
Yeah that was what I was suggesting re object envy - extracting registry to a configurable-at-instantiatiation immutable hash-quackalike would be the nice version, but feels like wasted effort and global complexity in this context.